conjure-undertow-lib/src/main/java/com/palantir/conjure/java/undertow/lib/SerializerRegistry.java

```diff
@@ -16,77 +16,16 @@
 
 package com.palantir.conjure.java.undertow.lib;
 
-import com.google.common.annotations.VisibleForTesting;
-import com.google.common.collect.ObjectArrays;
 import com.google.common.reflect.TypeToken;
-import com.palantir.logsafe.Preconditions;
-import com.palantir.logsafe.SafeArg;
-import com.palantir.logsafe.exceptions.SafeIllegalArgumentException;
 import io.undertow.server.HttpServerExchange;
-import io.undertow.util.HeaderValues;
-import io.undertow.util.Headers;
 import java.io.IOException;
 
 /** Orchestrates serialization and deserialization of response and request bodies. */
-public final class SerializerRegistry {
-
-    private final Serializer defaultSerializer;
-    private final Serializer[] serializers;
-
-    /**
-     * Creates a registry that {@link #getResponseSerializer selects} the first (based on input order) of the
-     * provided serializers that  {@link Serializer#supportsContentType supports} the serialization format
-     * {@link Headers#ACCEPT accepted} by a given request, or the first serializer if no such serializer can be found.
-     */
-    public SerializerRegistry(Serializer defaultSerializer, Serializer... serializers) {
-        this.defaultSerializer = defaultSerializer;
-        this.serializers = ObjectArrays.concat(defaultSerializer, serializers);
-    }
-
-    /** Returns the {@link Serializer} to use to deserialize the request body. */
-    @VisibleForTesting
-    Serializer getRequestDeserializer(HttpServerExchange exchange) {
-        String contentType = exchange.getRequestHeaders().getFirst(Headers.CONTENT_TYPE);
-        if (contentType == null) {
-            throw new SafeIllegalArgumentException("Request is missing Content-Type header");
-        }
-        for (Serializer serializer : serializers) {
-            if (serializer.supportsContentType(contentType)) {
-                return serializer;
-            }
-        }
-        throw new SafeIllegalArgumentException("Unsupported Content-Type",
-                SafeArg.of("Content-Type", contentType));
-    }
-
-    /** Returns the {@link Serializer} to use for the exchange response. */
-    @VisibleForTesting
-    Serializer getResponseSerializer(HttpServerExchange exchange) {
-        HeaderValues acceptValues = exchange.getRequestHeaders().get(Headers.ACCEPT);
-        if (acceptValues != null) {
-            // This implementation prefers the client "Accept" order
-            for (String acceptValue : acceptValues) {
-                for (Serializer serializer : serializers) {
-                    if (serializer.supportsContentType(acceptValue)) {
-                        return serializer;
-                    }
-                }
-            }
-        }
-        // Fall back to the default
-        return defaultSerializer;
-    }
+public interface SerializerRegistry {
```

Comment:
In the current formm as an interface, the methods make it more akin to a "HttpExchangeSerializer" than a registry.
Since the Serializer has been moved to the runtime library, one could argue that implementing it as a SerializerRegistry is an implementation detail of the runtime.

Reply:
I opted to avoid changing the generated code in this PR.

There are some changes I'd like to make to the `Serializer` and `SerializerRegistry` APIs to avoid blocking ourselves in a corner:

Serializer -> Encoding
SerializerRegistry -> SerializationContext

Update serializer and serializerregistry(context?) to create typed Serializer and Deserializer objects to avoid type info lookups per request

Reply:
Sounds good. If you're looking for inspiration for typesafe encoding/decoding, this library could be insightful, even just for the API: https://circe.github.io/circe/
Unfortunately, it relies on implicits which don't exist in Java.